### PR TITLE
Fix bug 1731483 (MEMORY storage engine incorrectly allows BLOB column…

### PR DESCRIPTION
**Pull request by laurynas-biveinis**
…s before indexed columns)

Dynamic row format feature to support BLOB/VARCHAR in MEMORY tables
requires all the key columns to come before any BLOB columns. This
requirement however was not enforced, allowing creating MEMORY tables
in unsupported column configurations, which then crash or lose data in
usage. Fix by
- reverting the previous fix for bug 1706356, which shares the root
cause;
- adding the check in heap_create whether there are any BLOB columns
before any key columns, and returning ER_TABLE_CANT_HANDLE_BLOB if
so.

http://10.30.6.254/view/PS%205.5/job/percona-server-5.5-param/1593/

### CODE REVIEW

File: mysql-test/t/percona_heap_var.test [hunk 1]
```diff
@@ -83,3 +84,63 @@ select count(*) from t1;
...
+
+CREATE TABLE t1 (a INT PRIMARY KEY, b BLOB, c INT, KEY c(c)) ENGINE=InnoDB;
+--error ER_TABLE_CANT_HANDLE_BLOB
+ALTER TABLE t1 ENGINE=MEMORY;
+DROP TABLE t1;
+
+CREATE TABLE t1 (a INT PRIMARY KEY, b TEXT, c INT, KEY c(c)) ENGINE=InnoDB;
+--error ER_TABLE_CANT_HANDLE_BLOB
+ALTER TABLE t1 ENGINE=MEMORY;
+DROP TABLE t1;
```
Comment (percona-ysorokin) [inline mysql-test/t/percona_heap_var.test:146]: I would also add checks for column reordering
```sql
CREATE TABLE t1 (a INT PRIMARY KEY, b BLOB) ENGINE=MEMORY;
--error ER_TABLE_CANT_HANDLE_BLOB
ALTER TABLE t1 MODIFY b BLOB FIRST;
```

File: mysql-test/r/percona_heap_var.result [hunk 2]
```diff
@@ -192,3 +192,50 @@ count(*)
 10001
 set @@session.max_heap_table_size=default;
 drop table t1;
+#
+# Bug 1731483: MEMORY storage engine incorrectly allows BLOB columns before indexed columns
+#
+CREATE TABLE t1 (a BLOB, b INT PRIMARY KEY) ENGINE=MEMORY;
+ERROR 42000: The used table type doesn't support BLOB/TEXT columns
```
Comment (robgolebiowski) [inline mysql-test/r/percona_heap_var.result:199]: Could error message be more specific ? From this I would assume it is not possible at all to create BLOB/TEXT columns inside memory table

File: mysql-test/t/percona_heap_var.test (hunk 1 above)
Comment (percona-ysorokin) [reply mysql-test/t/percona_heap_var.test:146]: And for composite keys:
```sql
--error ER_TABLE_CANT_HANDLE_BLOB
CREATE TABLE t1 (a INT, b BLOB, c INT, PRIMARY KEY(a, c)) ENGINE=MEMORY;

CREATE TABLE t1 (a INT, b BLOB, c INT) ENGINE=MEMORY;
--error ER_TABLE_CANT_HANDLE_BLOB
ALTER TABLE t1 ADD INDEX idx_composite (a, c);
--error ER_TABLE_CANT_HANDLE_BLOB
ALTER TABLE t1 ADD PRIMARY KEY (a, c);
```

**Review by robgolebiowski — CHANGES_REQUESTED** [review]

File: storage/heap/ha_heap.cc [hunk 3]
```diff
@@ -854,6 +854,15 @@ heap_prepare_hp_create_info(TABLE *table_arg, bool internal_table,
       {
         fixed_data_size= next_field_pos;
       }
+
+
+      if (field->field_index >= fixed_key_fieldnr)
```
Comment (percona-ysorokin) [inline storage/heap/ha_heap.cc:859]: Will this work for the case when BLOB is between fields being indexed, like
```sql
CREATE TABLE t1 (a INT, b BLOB, c INT, PRIMARY KEY(a, c)) ENGINE=MEMORY
```

File: mysql-test/r/percona_heap_var.result (hunk 2 above)
Comment (laurynas-biveinis) [reply mysql-test/r/percona_heap_var.result:199]: I have reused one of the existing messages, which is also used in the feature patch elsewhere. IMHO we should not add new error codes to 5.5 at this point

File: storage/heap/ha_heap.cc (hunk 3 above)
Comment (laurynas-biveinis) [reply storage/heap/ha_heap.cc:859]: Yes this code iterates through all the key parts. The testcase above for the composite keys covers it

**Review by robgolebiowski — APPROVED** [review]

**Review by percona-ysorokin — APPROVED** [review]
LGTM

**Comment by laurynas-biveinis** [comment]
http://jenkins.percona.com/job/percona-server-5.5-param/1594/

**Pull request merged**